File: benchmarks/text/PathNormalize.c
#include <stdint.h>
#include <stddef.h>

static int normalize_separators(char *destination, size_t capacity, const char *source, size_t *length) {
    size_t write = 0;
    int previous_separator = 0;

    for (size_t read = 0; source[read] != '\0'; read += 1) {
        char value = source[read] == '\\' ? '/' : source[read];
        int separator = value == '/';
        if (separator && previous_separator) {
            continue;
        }

        if (write + 1 >= capacity) {
            return 0;
        }

        destination[write] = value;
        write += 1;
        previous_separator = separator;
    }

    destination[write] = '\0';
    *length = write;
    return 1;
}

int main(void) {
    char buffer[64] = {0};
    int64_t checksum = 0;

    for (int32_t i = 0; i < 5000; i += 1) {
        size_t length = 0;
        if (!normalize_separators(buffer, sizeof(buffer), "alpha//beta///gamma.txt", &length)) {
            return 2;
        }

        checksum += (int64_t)length;

        if (!normalize_separators(buffer, sizeof(buffer), "alpha///beta.txt", &length)) {
            return 3;
        }

        checksum += (int64_t)length;
    }

    return checksum == 170000 ? 0 : 4;
}

File: benchmarks/text/PathNormalize.rs
fn normalize_separators(output: &mut String, source: &str) -> bool {
    output.clear();
    let mut previous_separator = false;

    for mut value in source.bytes() {
        if value == b'\\' {
            value = b'/';
        }

        let separator = value == b'/';
        if separator && previous_separator {
            continue;
        }

        if output.len() + 1 > output.capacity() {
            return false;
        }

        output.push(char::from(value));
        previous_separator = separator;
    }

    true
}

fn main() {
    let mut output = String::with_capacity(64);
    let mut checksum = 0_i64;

    for _ in 0_i32..5000 {
        if !normalize_separators(&mut output, "alpha//beta///gamma.txt") {
            std::process::exit(2);
        }

        checksum += output.len() as i64;

        if !normalize_separators(&mut output, "alpha///beta.txt") {
            std::process::exit(3);
        }

        checksum += output.len() as i64;
    }

    if checksum != 170_000 {
        std::process::exit(4);
    }
}
